src/sparse_ir/spr.py

```diff
@@ -1,66 +1,135 @@
 # Copyright (C) 2020-2021 Markus Wallerberger and others
 # SPDX-License-Identifier: MIT
 import numpy as np
+from .kernel import LogisticKernel, RegularizedBoseKernel
 from .basis import FiniteTempBasis
-from .sampling import DecomposedMatrix
+from .sampling import MatsubaraSampling, TauSampling, DecomposedMatrix
+from typing import Optional
+
+
+class MatsubaraPoleBasis:
+    def __init__(self, beta: float, poles: np.ndarray) -> None:
+        self._beta = beta
+        self._poles = poles
+
+    def __call__(self, n: np.ndarray) -> np.ndarray:
+        """Evaluate basis functions at given frequency n"""
+        iv = 1j*n * np.pi/self._beta
+        return 1/(iv[None,:] - self._poles[:,None])
+
+
+class TauPoleBasis:
+    def __init__(
+            self,
+            beta: float,
+            statistics: str,
+            poles: np.ndarray) -> None:
+        self._beta = beta
+        self._statistics = statistics
+        self._poles = poles
+        self._wmax = np.abs(poles).max()
+
+    def __call__(self, tau: np.ndarray) -> np.ndarray:
+        """ Evaluate basis functions at tau """
+        if (tau < 0).any() or (tau > self._beta).any():
+            raise RuntimeError("tau must be in [0, beta]!")
```

Comment:
This type annotation is overly restrictive IMHO.  You may very well call this function just with the value 0 rather than a numpy array.  This is why we need an explicit conversion also here:
```suggestion
    def __call__(self, tau) -> np.ndarray:
        """ Evaluate basis functions at tau """
        tau = np.asarray(tau)   
        if (tau < 0).any() or (tau > self._beta).any():
            raise RuntimeError("tau must be in [0, beta]!")
```